### ISSUE TITLE
reading content inside request handler blocks

### ISSUE THREAD

**Issue by david-j-peterson** [ReactiveX/RxNetty#347]
This contrived simple request handler will block forever:

```
public class ContentReader {
    public static void main(String[] args) throws Exception {
        ContentReader proxy = new ContentReader();
    }

    public ContentReader() throws Exception {
        Server.startAndWait();
    }

    private final HttpServer<ByteBuf, ByteBuf> Server = RxNetty.newHttpServerBuilder(8800, this::onRequest)
                                                               .pipelineConfigurator(new HttpServerPipelineConfigurator<>())
                                                               .enableWireLogging(LogLevel.ERROR).build();

    private Observable<Void> onRequest(final HttpServerRequest<ByteBuf> cRequest, final HttpServerResponse<ByteBuf> cResponse) {
        System.err.println("ContentReader => cRequest: " + cRequest.getPath());

        cRequest.getContent()
                .map(byteBuf -> {
                    System.err.println(byteBuf.toString(Charset.defaultCharset()));
                    return null;
                })
                .toBlocking()
                .last();

        return null;
    }
}
```

But changing it to:

```
    private Observable<Void> onRequest(final HttpServerRequest<ByteBuf> cRequest, final HttpServerResponse<ByteBuf> cResponse) {
        System.err.println("ContentReader => cRequest: " + cRequest.getPath());

        return cRequest.getContent()
                       .map(byteBuf -> {
                           System.err.println(byteBuf.toString(Charset.defaultCharset()));
                           return null;
                       });
    }
```

works - the POST'ed data will get printed out.

This may be related to an actual problem I'm having trying to write a transparent proxy using RxNetty. The following also seems to hang - or at least the submitted HttpClientRequest is never executed when there is POST'ed data.

```
    private Observable<Void> onRequest(final HttpServerRequest<ByteBuf> cRequest, final HttpServerResponse<ByteBuf> cResponse) {
        logger.info("RxProxy => cRequest: " + cRequest.getPath());

        HttpClientRequest<ByteBuf> pRequest = HttpClientRequest.create(HttpVersion.HTTP_1_1, cRequest.getHttpMethod(), cRequest.getUri());

        cRequest.getHeaders()
                .entries()
                .stream()
                .filter(entry -> !HopHeaders.contains(entry.getKey().toLowerCase()))
                .forEach(entry -> pRequest.withHeader(entry.getKey(), entry.getValue()));

        pRequest.withContentSource(cRequest.getContent());

        return Client.submit(pRequest)
                     .flatMap(pResponse -> {
                         cResponse.setStatus(pResponse.getStatus());

                         cResponse.getHeaders().clear();
                         pResponse.getHeaders()
                                  .entries()
                                  .stream()
                                  .filter(entry -> !HopHeaders.contains(entry.getKey().toLowerCase()))
                                  .forEach(entry -> cResponse.getHeaders().add(entry.getKey(), entry.getValue()));

                         return pResponse.getContent();
                     })
                     .flatMap(byteBuf -> {
                         cResponse.writeBytes(byteBuf.copy());
                         return Observable.just(cResponse);
                     })
                     .flatMap(HttpServerResponse::flush);
    }
```

However the above code does work to proxy GET requests.

It's possible that the Observable returned from cRequest.getContent() needs to be duplicated and/or the ByteBuf's need to be copied to a new Observable to be used as the content source for pRequest, but I can't figure out how to structure the code to process the getContent() Observable outside the request handler (to keep it from blocking) and then somehow append the client submit() call to that.

Any suggestions? Is the blocking behavior a bug, or is this all just user error?

**Comment by NiteshKant** [comment]
The first example (`ContentReader`) hangs because you are blocking the eventloop when you do this:

``` java
                .toBlocking()
                .last();
```

Since, the eventloop is blocked, you will never receive any more data on that channel (and any other channel served by that eventloop)

> It's possible that the Observable returned from cRequest.getContent() needs to be duplicated and/or the ByteBuf's need to be copied to a new Observable to be used as the content source for pRequest

Yes, the reason why the proxy code fails for requests with content is that because RxNetty expects users to eagerly subscribe to the content. In the proxy case, since the content `Observable` is subscribed when the client establishes the connection and starts to write the response, that delayed subscription throws an error (post a timeout, so the behavior isn't consistent). In order to get around this problem, we have taken an approach of eagerly subscribing to the content and caching it in a `Subject` which then gets fed to the client as the request content. You can see how this is done in [Zuul 2.0](https://github.com/Netflix/zuul/tree/2.x) [here](https://github.com/Netflix/zuul/blob/2.x/zuul-core/src/main/java/com/netflix/zuul/lifecycle/FilterProcessor.java#L81)

Having said the above, this behavior is going to be fixed in [0.5.0](https://github.com/ReactiveX/RxNetty/milestones/0.5.0) by introducing back-pressure and on-demand read from channel (issue #324 ).

**Comment by NiteshKant** [comment]
@nrdavid I am closing this, if you think that this is a bug, reopen.
